include/solvers/PreconditionCG.hpp

```diff
@@ -0,0 +1,256 @@
+                }
+
+
+                M = (D + L) * inv(D) * (D + U);
+            }
+
+            if (type.compare("SOR preconditioning")==0){
+                double omega = 1.2;  // omega is in the range of (0,2) to make sure converge
+                std::size_t n = A.columns();
+                MatrixType L(n,n), D(n,n);
+                for(std::size_t i = 0; i < n; ++i){
+                    for(std::size_t j = 0; j < n; ++j){
+                        if(j < i){
+                            L(i,j) = A(i,j);
+                        }
+                        else{
+                            L(i,j) = 0;
+                        }
+                    }
+                }
+
+                for(std::size_t i = 0; i < n; ++i){
+                    for(std::size_t j = 0; j < n; ++j){
+                        if(j == i){
+                            D(i,j) = A(i,j);
+                        }
+                        else {
+                            D(i,j) = 0;
+                        }
+                    }
+                }
+
+                M = (D + omega * L)/omega;
+            }
+
+            if (type.compare("SSOR preconditioning")==0){
+                std::size_t n = A.columns();
+                MatrixType L(n,n), D(n,n);
+                for(std::size_t i = 0; i < n; ++i){
+                    for(std::size_t j = 0; j < n; ++j){
+                        if(j < i){
+                            L(i,j) = A(i,j);
+                        }
+                        else{
+                            L(i,j) = 0;
+                        }
+                    }
+                }
+
+                for(std::size_t i = 0; i < n; ++i){
+                    for(std::size_t j = 0; j < n; ++j){
+                        if(j == i){
+                            D(i,j) = A(i,j);
+                        }
+                        else {
+                            D(i,j) = 0;
+                        }
+                    }
+                }
+
+                M = (D + L) * inv(D) * trans(D + L);
+            }
+
+
+            if (type.compare("incomplete Cholesky factorization") == 0 || type.compare("") == 0){
+                // The Cholesky factorization of A is A = LL*, where L is a lower triangular matrix.
+                // Incomplete Cholesky factorization precondition is M =KK*, where a sparse lower triangular matrix K is close to L.
+                // Finding the exact Cholesky decomposition, except that any entry is set to zero
+                // if the corresponding entry in A is also zero.
+
+                MatrixType L, K;
+                llh( A, L );  // L* LH decomposition of a row-major matrix
+                K = L;
+
+                std::size_t n = A.columns();
+                for(std::size_t i = 0; i < n; ++i){
+                    for(std::size_t j = 0; j < n; ++j){
+                        if (A(i, j) == 0)
+                            K(i, j) = 0;
+                    }
+                }
```

Comment:
this could take advantage of `blaze::map` (https://bitbucket.org/blaze-lib/blaze/wiki/Matrix%20Operations).

Reply:
I don't quit understand how to take advantage of blaze::map, since blaze::map is used to apply a custom operation on each element of matrix A. 

Our goal is to change some elements to zero (which are zeros in the same locations of matrix A) in the lower matrix L,  which is obtained after Cholesky factorization. 

For example,
![1483371663](https://user-images.githubusercontent.com/29106484/61083244-da369700-a3f0-11e9-98ef-d893b4adca16.jpg)

Reply:
One possible solution might be: 
 First define a new matrix B, every elements in B is one, except those who have the same locations of the zero elements in A. Then do element-wise between matrix B and matrix L to get matrix K.

Reply:
You could instead of:
```
for(std::size_t i = 0; i < n; ++i){ 
    for(std::size_t j = 0; j < n; ++j){ 
        if (A(i, j) == 0) 
            K(i, j) = 0; 
     } 
}
```
you could write:
```
K = blaze::map(A, L, [](T aval, T lval) { return aval == T{0} ? T{0} : lval; });
```

Reply:
Yes! Thanks. It works well and the result is correct.